pkg/testutils/fullstack.go

```diff
@@ -0,0 +1,213 @@
+package testutils
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"database/sql"
+	"encoding/hex"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind/backends"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/stretchr/testify/require"
+	"github.com/textileio/go-tableland/internal/chains"
+	"github.com/textileio/go-tableland/internal/router"
+	"github.com/textileio/go-tableland/internal/system"
+	systemimpl "github.com/textileio/go-tableland/internal/system/impl"
+	"github.com/textileio/go-tableland/internal/tableland"
+	"github.com/textileio/go-tableland/internal/tableland/impl"
+	"github.com/textileio/go-tableland/pkg/client"
+	"github.com/textileio/go-tableland/pkg/eventprocessor/eventfeed"
+	efimpl "github.com/textileio/go-tableland/pkg/eventprocessor/eventfeed/impl"
+	epimpl "github.com/textileio/go-tableland/pkg/eventprocessor/impl"
+	executor "github.com/textileio/go-tableland/pkg/eventprocessor/impl/executor/impl"
+	nonceimpl "github.com/textileio/go-tableland/pkg/nonce/impl"
+	"github.com/textileio/go-tableland/pkg/parsing"
+	parserimpl "github.com/textileio/go-tableland/pkg/parsing/impl"
+	"github.com/textileio/go-tableland/pkg/sqlstore"
+	sqlstoreimpl "github.com/textileio/go-tableland/pkg/sqlstore/impl"
+	sqlstoreimplsystem "github.com/textileio/go-tableland/pkg/sqlstore/impl/system"
+	sqlstoreimpluser "github.com/textileio/go-tableland/pkg/sqlstore/impl/user"
+	"github.com/textileio/go-tableland/pkg/tables"
+	"github.com/textileio/go-tableland/pkg/tables/impl/ethereum"
+	"github.com/textileio/go-tableland/pkg/tables/impl/testutil"
+	"github.com/textileio/go-tableland/pkg/wallet"
+	"github.com/textileio/go-tableland/tests"
+)
+
+// ChainID is the test chain id.
+var ChainID = tableland.ChainID(1337)
+
+// FullStack holds all potentially useful components of the Tableland test stack.
+type FullStack struct {
+	Backend           *backends.SimulatedBackend
+	Address           common.Address
+	Contract          *ethereum.Contract
+	TransactOpts      *bind.TransactOpts
+	Sk                *ecdsa.PrivateKey
+	Wallet            *wallet.Wallet
+	TblContractClient *ethereum.Client
+	Client            *client.Client
+}
+
+// Deps holds possile dependencies that can optionally be provided to spin up the full stack.
+type Deps struct {
+	DBURI         string
+	Parser        parsing.SQLValidator
+	SystemStore   sqlstore.SystemStore
+	UserStore     sqlstore.UserStore
+	ACL           tableland.ACL
+	Tableland     tableland.Tableland
+	SystemService system.SystemService
+}
+
+// CreateFullStack creates the full stack, returning possibly useful things.
+func CreateFullStack(t *testing.T, deps Deps) FullStack {
```

Comment:
The goal here is to provide a general use full stack test harness that also allows you to provide custom implementations or mocks for any of the important dependencies that are defined as `interface` types. If you don't provide a particular dependency via the `deps Deps` arg, a default is used.

There is a "user beware" situation here that I couldn't find a good way around: If you provide a dependency A that was initialized with another instance B that is also represented by `Deps`, you need to pass that instance B in as part of `deps Deps` to make sure the entire dependency tree is consistent (for example, to avoid a situation where `Tableland` is using a mocked version of `UserStore`, but `SystemService` is using the default `UserStore`).

Reply:
Yeah, we have the problem of instantiating dependencies for testing all over the repo. And each place does on it's one way. I think this is good. I can see and future where we have a library with a flexible nice API used in every test (maybe even in main) to create these subcomponents. This can be the beginning of such a library.